Support for Memory Statisticsd Process and Configurations

Arham-Nasir opened the pull request:
Why I did it:
The Memory_Statisticsd Process was added to enhance system monitoring by collecting memory usage data at configurable intervals. This will allow the system to store historical data, making it easier to analyze memory trends and debug memory-related issues.

Arham-Nasir commented:
Hi @qiluo-msft  , @prgeor  , and @FengPan-Frank ,

Could you please help review the HLD and the linked code PRs for our feature?

Thank you!

FengPan-Frank commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,232 @@
+                memory_statistics = self.collect_memory_statistics()
+                self.store_memory_statistics(memory_statistics)
+            except Exception as e:
+                self.logger.log(f"Error collecting or storing memory statistics: {e}", logging.ERROR)
```
will this be repeated in dead loop and generate bunch of logs?

Arham-Nasir replied:
Thank you for your feedback. The memory statistics daemon is designed with robust loop control to prevent dead loops. It collects data at a configurable interval, logs it in compressed format for historical data, and shuts down gracefully when triggered. The loop control mechanism ensures it waits according to the sampling interval and only continues running until a shutdown signal is received. Potential issues could arise from misconfigurations (e.g., zero or negative sampling intervals) but the daemon is designed to handle such cases effectively.

ridahanif96 commented:
@qiluo-msft @xincunli-sonic pls help review this feature

Arham-Nasir commented:
> > Hi @FengPan-Frank and @qiluo-msft , When you have a moment, could you please review this PR? I would appreciate your feedback and suggestions. Thank you for your time and guidance
> 
> Thanks for the work, could you add test case to cover collector service code?

Thank you for the review and suggestion, I’ll add test cases for the collector service code. Please let me know if there are any specific scenarios you'd like me to cover.

FengPan-Frank commented:
> > > Hi @FengPan-Frank and @qiluo-msft , When you have a moment, could you please review this PR? I would appreciate your feedback and suggestions. Thank you for your time and guidance
> > 
> > 
> > Thanks for the work, could you add test case to cover collector service code?
> 
> Thank you for the review and suggestion, I’ll add test cases for the collector service code. Please let me know if there are any specific scenarios you'd like me to cover.

Just common sanity check to cover the daemon functions.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics.conf:
```diff
@@ -0,0 +1,10 @@
+# Memory Statistics Daemon Configuration
```
Would it better if these configurations can put into config db instead of a static conf file?

Arham-Nasir replied:
We have stored configurations in both config_db and a static conf file for flexibility and reliability. 

At startup, the daemon reads default settings from the conf file to ensure a consistent state after a restart. During runtime, the daemon dynamically updates its configuration from config_db without needing a restart. This approach combines the stability of predefined defaults with the convenience of live updates, ensuring predictable behavior and seamless adaptability.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+#!/usr/bin/env python3
+import psutil
```
Can you sort import as nit purpose?

Arham-Nasir replied:
Thank you for the review. I have organized the imports following the standard Python import grouping practice: standard library imports, third-party imports, and local application imports, sorted alphabetically within each group for improved readability and maintainability.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+        """Logs a message with the 'DEBUG' level."""
+        self.log(syslog.LOG_DEBUG, message)
+
+    def close_logger(self):
```
>close_logger [](http://example.com/codeflow?start=8&length=12)

There is no caller for this close method?

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+                        logger.log_info(f"Removing outdated entry: {old_entry}")
+                        del total_dict['system_memory'][memory_type][old_entry]
+
+        total_dict['system_memory']['count'] = len(total_dict['system_memory'].get('system', {}))
```
>count [](http://example.com/codeflow?start=37&length=5)

A constant?

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+                logger.log_info("Accepted new connection")
+                self.handle_connection(connection)
+            except socket.timeout:
+                continue  
```
>continue [](http://example.com/codeflow?start=16&length=8)

Add log in case timeout happen, but lost debuggability.

Arham-Nasir replied:
Thanks for the feedback! I've added a logger.log_debug for socket.timeout to improve debuggability.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+        """
+        self.pid_file = pid_file
+
+    def daemonize(self):
```
>daemonize [](http://example.com/codeflow?start=8&length=9)

Add validation to confirm the daemonization process succeeded and log an appropriate error if it did not.

Arham-Nasir replied:
Hi @xincunli-sonic ,

The enhanced Daemonizer class is designed to ensure reliable daemon creation, with robust validation for process isolation, session management, and file descriptor redirection. 

Thank you.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+            if slot < num_columns:
+                self.add_entry_to_time_group_list(time_entry_summary, slot, memory_entry)
+
+    def aggregate_data(self, request_data, time_entry_summary, num_columns):
```
>aggregate_data [](http://example.com/codeflow?start=8&length=14)

In aggregate_data, the logic for adding entries to the time_group_list could be inefficient for large datasets due to multiple nested calls and list manipulations.
Suggestion: Use batch processing or caching mechanisms to reduce redundant operations and improve performance.

Arham-Nasir replied:
Thank you for your valuable feedback and suggestion to optimize the aggregate_data function. I have updated the code by implementing batch processing with dynamic sizing to reduce redundant operations, minimizing nested calls for better performance, and adding robust error handling with detailed logging. Let me know if further adjustments are needed.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+        logger.log_info(f"Service initialized with name: {self.name}")
+
+        self.config_file_path = config_file_path
+        self.memory_statistics_lock = threading.Lock()
```
>memory_statistics_lock [](http://example.com/codeflow?start=13&length=22)

Please check all code for shared resource access and ensure that locks are applied consistently, especially when modifying global state.

Arham-Nasir replied:
Thank you for pointing this out. I have carefully reviewed the code and ensured that locks are applied consistently for all shared resource access.

Added ThreadSafeConfig class for safe configuration access and consistently used memory_statistics_lock during memory statistics collection to prevent race conditions. 

Let me know if further adjustments are needed

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+        self.memory_collection_thread = None
+
+        self.config = memory_statistics_config.copy()
+        self.load_config_from_file()
```
>load_config_from_file [](http://example.com/codeflow?start=13&length=21)

If config load failed, are we going to proceed further?

Arham-Nasir replied:
Yes, if an error occurs while loading the configuration file (e.g., due to a parsing error or file corruption), the service logs a warning message and falls back to default settings. This ensures the service proceeds with a known, consistent state instead of halting. 

The benefit of this approach is that it minimizes downtime and allows the service to continue operating with reliable defaults to prevent unpredictable behavior, even if the configuration file is invalid or unavailable.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+        return Dict2Obj(cls._format_timedelta_as_dict(time_delta))
+    
+    @staticmethod
+    def format_memory_size(size):
```
>size [](http://example.com/codeflow?start=27&length=4)

Validate the input type and range, and add exception handling for invalid inputs.

Arham-Nasir replied:
Thank you for pointing this out. I've updated the format_memory_size method with comprehensive input validation, including type checking, range validation, and error handling as suggested. Please let me know if further adjustments are needed.

xincunli-sonic commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+            sys.exit(1)
+
+
+class MemoryStatisticsService:
```
>MemoryStatisticsService [](http://example.com/codeflow?start=6&length=23)

The MemoryStatisticsService class contains many responsibilities, such as socket handling, signal processing, and memory statistics collection, making it harder to maintain and test.

Suggestion: Split this class into smaller, more focused components to adhere to the single-responsibility principle.

Arham-Nasir commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,1877 @@
+        current_time = datetime.now()
+        retention_threshold = timedelta(days=self.retention_period)
+
+        if 'system_memory' in total_dict:
```
Thank you for the suggestion! I have addressed this by defining 'system_memory’ as a constant SYSTEM_MEMORY_KEY to improve clarity and maintainability.

FengPan-Frank reviewed (dismissed):
LGTM

mssonicbld commented:
/azp run Azure.sonic-buildimage

wajahatrazi commented:
Hi @qiluo-msft, 

Please help approve and merge.

zhangyanzhao commented:
@Arham-Nasir  @liushilongbuaa  can you please check the Semgrep failure? Thanks.

liushilongbuaa commented:
> @Arham-Nasir @liushilongbuaa can you please check the Semgrep failure? Thanks.

The failure is related with this PR.
```
    files/image_config/memory_statistics/memory_statistics_handler.py
    ❯❱ python.lang.security.deserialization.pickle.avoid-pickle
          Avoid using `pickle`, which is known to lead to code execution vulnerabilities. When unpickling, the
          serialized data could be manipulated to run arbitrary code. Instead, consider serializing the       
          relevant data as JSON or a similar text-based serialization format.                                 
          Details: https://sg.run/OPwB                                                                        
                                                                                                              
          640┆ loaded_entry = pickle.load(bfile)
            ⋮┆----------------------------------------
          776┆ pickle.dump(total_dict, bfile)
            ⋮┆----------------------------------------
          783┆ pickle.dump(mem_dict, bfile)
            ⋮┆----------------------------------------
          1138┆ memory_entry = pickle.load(bfile)
```

mssonicbld commented:
/azp run Azure.sonic-buildimage

Arham-Nasir commented:
> @Arham-Nasir @liushilongbuaa can you please check the Semgrep failure? Thanks.

Thank you for your review, @zhangyanzhao   and @liushilongbuaa  . The Semgrep warning issue was due to the use of the pickle library, which, while optimized for performance, may introduce security risks during deserialization.

To address this, I have replaced all pickle usage with secure JSON serialization, incorporated gzip compression for storage efficiency, and implemented robust error handling. These updates enhance security while preserving functionality. I kindly request your feedback on the changes and welcome any suggestions for further improvements.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,2139 @@
+                    raise ValueError(f"Could not parse date: {request[time_key]}")
+                formatted_date = parsed_date.strftime(date_format)
+            except Exception as e:
+                raise Exception(f"Date format error, input: {request[time_key]}, error: {str(e)}")
```
>Exception [](http://example.com/codeflow?start=22&length=9)

Do not raise generic exception type.

Arham-Nasir replied:
Thanks for your feedback. I have updated the error handling by replacing Exception with ValueError. Let me know if you have any further suggestions.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,2139 @@
+        Raises:
+        - ValueError: If the sampling interval is not within the accepted range.
+        """
+        hourly_rate = 60 / int(interval_minutes)
```
>/ [](http://example.com/codeflow?start=25&length=1)

Is there risk of dividing by 0?

Arham-Nasir replied:
There is no risk of division by zero since validation is already handled in the CLI and configuration part. However, I’ve added an extra safeguard to enhance robustness.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,2139 @@
+                        )
+                except json.JSONDecodeError as e:
+                    logger.log_error(f"Error decoding JSON from {file_name}: {e}")
+        except Exception as e:
```
>Exception [](http://example.com/codeflow?start=15&length=9)

It swallows exceptions, could you remove?

Arham-Nasir replied:
Thank you for Pointing this out. I’ve removed the broad Exception catch to ensure proper error propagation.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,2139 @@
+                
+                time_group_list[start:end] = processed_batch
+            
+            except Exception as e:
```
>Exception [](http://example.com/codeflow?start=19&length=9)

Could you use more specific exception type?

Arham-Nasir replied:
I have replaced the generic Exception with KeyError, IndexError, TypeError, ValueError to improve clarity in error handling. Thanks.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_handler.py:
```diff
@@ -0,0 +1,2139 @@
+#!/usr/bin/env python3
```
>python3 [](http://example.com/codeflow?start=15&length=7)

Could you write unit test? The line coverage should be >= 80%.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

wajahatrazi commented:
@qiluo-msft / @FengPan-Frank 

Folks please review, and help merge.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_service.py:
```diff
@@ -0,0 +1,2134 @@
+            if socket_address and os.path.exists(socket_address):
+                os.unlink(socket_address)
+                logger.log_info(f"Removed socket file: {socket_address}")
+        except Exception as e:
```
>except Exception as [](http://example.com/codeflow?start=8&length=19)

still too general Exception, still swallow exception.

Arham-Nasir replied:
Thank you for the feedback.
I have replaced except Exception with more specific exceptions to ensure better debugging and logging.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_service.py:
```diff
@@ -0,0 +1,2134 @@
+            if os.path.exists(pid_file):
+                os.unlink(pid_file)
+                logger.log_info(f"Removed PID file: {pid_file}")
+        except Exception as e:
```
>except Exception as [](http://example.com/codeflow?start=8&length=19)

The same

Arham-Nasir replied:
Replaced except Exception with specific exceptions for improved clarity and control.

qiluo-msft commented on files/image_config/memory_statistics/memory_statistics_service.py:
```diff
@@ -0,0 +1,2134 @@
+        try:
+            with gzip.open(memory_statistics_config['TOTAL_MEMORY_STATISTICS_LOG_FILENAME'], 'wt', encoding='utf-8') as jfile:
+                json.dump(total_dict, jfile)
+        except Exception as e:
```
>except Exception [](http://example.com/codeflow?start=8&length=16)

still too general Exception

Arham-Nasir replied:
Thank you for the feedback.
I’ve replaced all instances of the generic except Exception with more specific exceptions to ensure better error handling and clarity. Please let me know if you have any further suggestions.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

lguohan commented:
need to rethink the approach, this looks like reinvent the wheel of time series database on the sonic itself.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

Arham-Nasir commented:
> need to rethink the approach, this looks like reinvent the wheel of time series database on the sonic itself.

We want to clarify that the purpose of memorystatsd is not to build a time-series database (TSDB) inside the SONiC Redis DB. Instead, memorystatsd is designed to be a simple and lightweight solution that fits well in SONiC environments.
 
To keep the system efficient and reduce complexity, memorystatsd avoids using Redis for storing data. 
 
Instead, it saves compressed JSON snapshots to /var/log/memorystats/memory-stats.json.gz. 
 
Each snapshot is very compact **approximately 225 bytes** resulting in a total of around 63 KB per day. Depending on the configured retention period, the storage footprint remains minimal.
 
Old data is automatically deleted based on the set retention period, helping to meet platform limits and keep resource usage low.

wajahatrazi commented:
Dear @qiluo-msft 

This is the only PR left to be merged as per the scope of the HLD, hoping for your support.
